packages/react-client/src/hooks/devices.ts

```diff
@@ -1,33 +1,98 @@
 import { useFishjamContext } from "./fishjamContext";
 import type { Device, AudioDevice } from "../types";
+import { useVideoDeviceManager } from "./deviceManagers/videoDeviceManager";
+import { useAudioDeviceManager } from "./deviceManagers/audioDeviceManager";
+import { useCallback, useEffect, useRef } from "react";
+import { getAvailableMedia, getCorrectedResult } from "../mediaInitializer";
 
 export function useCamera(): Device {
   const {
-    state,
     videoTrackManager: { currentTrack, ...trackManager },
   } = useFishjamContext();
 
+  const { deviceState } = useVideoDeviceManager();
+
   const streamedTrackId = currentTrack?.trackId ?? null;
   const streamedTrack = currentTrack?.track ?? null;
   const stream = currentTrack?.stream ?? null;
-  const devices = state.devices.camera.devices ?? [];
-  const activeDevice = state.devices.camera.deviceInfo;
+  const devices = deviceState.devices ?? [];
+  const activeDevice = deviceState.media?.deviceInfo ?? null;
 
   return { ...trackManager, streamedTrack, streamedTrackId, stream, devices, activeDevice };
 }
 
 export function useMicrophone(): AudioDevice {
   const {
-    state,
     audioTrackManager: { currentTrack, ...trackManager },
   } = useFishjamContext();
 
+  const { deviceState } = useAudioDeviceManager();
+
   const isAudioPlaying = currentTrack?.vadStatus === "speech";
   const streamedTrackId = currentTrack?.trackId ?? null;
   const streamedTrack = currentTrack?.track ?? null;
   const stream = currentTrack?.stream ?? null;
-  const devices = state.devices.microphone.devices ?? [];
-  const activeDevice = state.devices.microphone.deviceInfo;
+  const devices = deviceState.devices ?? [];
+  const activeDevice = deviceState.media?.deviceInfo ?? null;
 
   return { ...trackManager, streamedTrack, streamedTrackId, isAudioPlaying, stream, devices, activeDevice };
 }
+
+type InitializeDevicesProps = { autoInitialize?: boolean };
+
+export const useInitializeDevices = (props?: InitializeDevicesProps) => {
+  const areDevicesInitialized = useRef(false);
```

Comment:
This hook works similar to just deleted `useSetupMedia`. The user can use it in many different components, and by doing so, the device manager will start multiple times, which I don't think we are able to handle. I think `areDevicesInitialized` should be part of a shared state, or `useInitializeDevices` shouldn't be a hook at all, just a simple function (in that case, areDevicesInitialized also needs to be in shared state).

Reply:
My bad, the ref should be kept in the provider in this case. thanks